test/jdk/java/net/Socket/ConnectFailTest.java

```diff
@@ -0,0 +1,268 @@
+ * questions.
+ */
+
+import jdk.test.lib.Utils;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.function.Executable;
+import org.junit.jupiter.api.function.ThrowingConsumer;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.nio.channels.AlreadyConnectedException;
+import java.nio.channels.SocketChannel;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/*
+ * @test
+ * @bug 8343791
+ * @summary verifies the socket state after `connect()` failures
+ * @library /test/lib
+ * @run junit/othervm --add-opens java.base/java.net=ALL-UNNAMED ConnectFailTest
+ */
+class ConnectFailTest {
+
+    private static final int DEAD_SERVER_PORT = 0xDEAD;
+
+    private static final InetSocketAddress REFUSING_SOCKET_ADDRESS = Utils.refusingEndpoint();
+
+    private static final InetSocketAddress UNRESOLVED_ADDRESS =
+            InetSocketAddress.createUnresolved("no.such.host", DEAD_SERVER_PORT);
+
+    @Test
+    void testUnresolvedAddress() {
+        assertTrue(UNRESOLVED_ADDRESS.isUnresolved());
+    }
+
+    /**
+     * Verifies socket is closed when {@code unboundSocket.connect()} fails.
+     * @param socket an unbound socket
+     */
+    @ParameterizedTest
+    @MethodSource("boundSockets")
+    void testBoundSocket(Socket socket) throws IOException {
+        try (socket) {
+            assertTrue(socket.isBound());
+            assertFalse(socket.isConnected());
+            assertThrows(IOException.class, () -> socket.connect(REFUSING_SOCKET_ADDRESS));
+            assertTrue(socket.isClosed());
+        }
+    }
+
+    @Test
+    void testConnectedSocket() throws Throwable {
+        testConnectedSocket(
+                ConnectFailTest::createConnectedSocket,
+                executable -> {
+                    SocketException exception = assertThrows(SocketException.class, executable);
+                    assertEquals("already connected", exception.getMessage());
+                });
+    }
+
+    @Test
+    void testConnectedNioSocket() throws Throwable {
+        testConnectedSocket(
+                ConnectFailTest::createConnectedNioSocket,
+                executable -> assertThrows(AlreadyConnectedException.class, executable));
+    }
```

Comment:
I assume that testConnectedSocket and testConnectedNioSocket could be @ParameterizedTest so it's called with a new created Socket.

Reply:
It would be called twice with a new created Socket. Socket.connect is specified to throw IOException if already connected, we shouldn't be throwing (or testing for) AlreadyConnectedException here. It may be that Net.translateToSocketException is missing that mapping (we should check that).

Reply:
I need to verify the thrown exception in the test:

* If the socket passed created using `Socket::new`, a `SocketException("already connected")` is thrown
* If the socket passed created using `SocketChannel.open().socket()`, an `AlreadyConnectedException` is thrown

AFAICT, the verifier needs to be passed along with the socket, and this implies either separate tests, which is what the current code does, or a separate test method source as in

```java
static List<Arguments> connectedSocketFactoriesAndReconnectFailureVerifiers() {
    return List.of(
            Arguments.of(
                    (Function<SocketAddress, Socket>) ConnectFailTest::createConnectedSocket,
                    (Consumer<Executable>) executable -> {
                        SocketException exception = assertThrows(SocketException.class, executable);
                        assertEquals("already connected", exception.getMessage());
                    }),
            Arguments.of(
                    (Function<SocketAddress, Socket>) ConnectFailTest::createConnectedNioSocket,
                    (Consumer<Executable>) executable -> assertThrows(AlreadyConnectedException.class, executable))
    );
}
```

Am I mistaken?

Reply:
I checked the socket adaptor and there is needed a bug here. SocketChannel throws AlreadyConnectedException and the socket adaptor should map that to SocketException("already connected"). It doesn't do that so we should fix it.

Reply:
I will push an enhancement here. Allow me some time, please.

Reply:
> I will push an enhancement here. Allow me some time, please.

That's fine, it seems we have lacked good tests for these exceptional cases so probably not surprising that some issues turn up.

Reply:
For the record, `Net::translateToSocketException` doesn't map `AlreadyConnectedException`.

Reply:
> I checked the socket adaptor and there is needed a bug here. SocketChannel throws AlreadyConnectedException and the socket adaptor should map that to SocketException("already connected"). It doesn't do that so we should fix it.

@AlanBateman, I presume you want me to fix that mapping in a separate JBS+PR, right?

Reply:
If Daniel is okay with it, then it might be easier to fix it here as part of this PR because the changes to add the mapping are touching exactly the focus of this PR, and the new test is testing the condition that exposes this.

Reply:
@AlanBateman, I've updated `Net` and `ConnectFailTest`. Would you mind checking them, please? (In the meantime, I will start a `tier1..3` run and update the CSR.)